keps/sig-api-machinery/NNNN-crd-compatibility-checking/README.md

```diff
@@ -0,0 +1,723 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+CRD schemas can be changed in backward incompatible ways. While constraining
+updates to CRD schemas to backward compatible changes is encouraged, it is
+not enforced.
+
+This KEP proposes introducing a `compatibilityChecking` field to Custom
+Resource Definitions that can be either `BackwardCompatible` or `Off`.
+When the field is set to `BackwardCompatible` any updates to the CRD that
+attempt to make a backward incompatible changes to the schema will be
+rejected.
+
+## Motivation
+
+Lack of compatibility checking is a risk to CRD developers since they
+might accidently make a backward compatible change without realizing their
+mistake until too late.
+
+It is also a risk to cluster operators. Backward incompatible changes to CRDs
+can result in control plane outages, data loss and misbehaved controllers.
+
+### Goals
+
+Introduce an option for enabling compatibility checking on CRDs that prevents
+statically enforcable backward incompatible changes from being made to the
+schema.
+
+### Non-Goals
+
+Detect all possible backward incompatible changes. For example, if the description
+of a CRD field is updated to recommend the field's contents be used in a different
+way than it was previouisly, the change may be semantically backward incompatible.
+But since the properties of the field are otherwise unchanged in the schema, we
+cannot enforce backward compatibility with the proposed compatibility checker.
+
+Enforce compatibility between API versions. For example, if a CRD introduces `v2`
+but does not properly handle conversion between the existing `v1` and the new `v2`
+versions, the incompatibility will not be detected by this proposed compatibility
+checker.
+
+## Proposal
+
+Add a `spec.compatibilityChecking` enum to Custom Resource Definitions with the allowed
+values of:
+
+- `Off` (default)
+- `BackwardCompatible`
+
+`BackwardCompatible` will enforce:
+
+- Field types cannot be changed
+- Field names cannot be changed
+- Fields cannot be removed (but can be documented as deprecated)
+- Only optional or defaulted fields can be added (required fields cannot be added)
+- Enums (and Unions if/when they are supported):
+  - Enumeratable values cannot be removed
+  - Enumeratable values can only be added if the enum is optional or has a default? (TODO: reason this through fully)
```

Comment:
probably maps to guidance in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field, but I'm not sure how to determine that programmatically

> Older versions of expected API clients must be able handle the new value in a safe way:
>
> * If the enum field drives behavior of a single component, ensure all versions of that component that will encounter API objects containing the new value handle it properly or fail safe. For example, a new allowed value in a Pod enum field consumed by the kubelet must be handled safely by kubelets up to two versions older than the first API server release that allowed the new value.
>* If an API drives behavior that is implemented by external clients (like Ingress or NetworkPolicy), the enum field must explicitly indicate that additional values may be allowed in the future, and define how unrecognized values must be handled by clients. If this was not done in the first release containing the enum field, it is not safe to add new values that can break existing clients.